### PR TITLE
Client 2

### PR DESCRIPTION
**Pull request by valhristov**
What's changed:
- the requests are no longer in a single class, but are in separate classes for each request
- the request/response messages are built into the request classes because they can be versioned too (properties, structure, etc. could change)
- each request class is registered into the RequestFactory, which takes care about versioning and what request to execute, e.g. no need for if-s in the code. If you want to handle a breaking change, either create a new request, or override an existing.
- the requests return models, each request handles errors and http response codes (because this could also change between the api versions)
- ~RequestGenerator is the testing app that runs the service methods against the specified SQ instance~
- did not write the task cancellation tests, not sure if there is any benefit from them, but should be easy enough

### CODE REVIEW

File: src/SonarQube.Client.RequestGenerator/ArgsParser.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+﻿using System.Linq;
+
+namespace SonarQube.Client.RequestGenerator
+{
+    public class ArgsParser
```
Comment (valhristov) [inline src/SonarQube.Client.RequestGenerator/ArgsParser.cs]: That's a super simple cmd line args parser that provides default values for missing arguments

File: src/SonarQube.Client.RequestGenerator/Program.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        static void Main(string[] args)
+        {
+            var requestFactory = new RequestFactory();
+            DefaultConfiguration.Configure(requestFactory);
+
+            var cts = new CancellationTokenSource();
+
+            var service = new SonarQubeService2(new HttpClientHandler(), requestFactory);
+            //var service = new SonarQubeService(new SonarQubeClientFactory());
```
Comment (SonarTech) [inline src/SonarQube.Client.RequestGenerator/Program.cs]: MAJOR Remove this commented out code. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS125)

File: src/SonarQube.Client.RequestGenerator/Program.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Net.Http;
+using System.Threading;
+using SonarQube.Client.Api.Requests;
+using SonarQube.Client.Services;
+
+namespace SonarQube.Client.RequestGenerator
+{
+    class Program
```
Comment (SonarTech) [inline src/SonarQube.Client.RequestGenerator/Program.cs]: MAJOR Add a 'protected' constructor or the 'static' keyword to the class declaration. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1118)

File: src/SonarQube.Client/Services/SonarQubeService2.cs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        {
+            var qualityProfiles = await InvokeRequestAsync<IGetQualityProfilesRequest, SonarQubeQualityProfile[]>(
+                request =>
+                {
+                    request.ProjectKey = projectKey;
+                    request.OrganizationKey = organizationKey;
+                },
+                token);
+
+            // TODO: add language to the request configuration above
```
Comment (SonarTech) [inline src/SonarQube.Client/Services/SonarQubeService2.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/SonarQube.Client/Api/Requests/V5_20/GetRoslynExportProfileRequest.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        [JsonProperty("exporterKey", Required = Required.Always)]
+        public string ExporterKey => $"roslyn-{LanguageKey}";
+
+        protected override string Path => "api/qualityprofiles/export";
+
+        protected override RoslynExportProfileResponse ParseResponse(string response)
+        {
+            using (var reader = new StringReader(response))
+            {
+                // TODO: consider not returning the xml directly
```
Comment (SonarTech) [inline src/SonarQube.Client/Api/Requests/V5_20/GetRoslynExportProfileRequest.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/SonarQube.Client/Api/Requests/V6_30/GetPropertiesRequest.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+        private class PropertyResponse
+        {
+            [JsonProperty("key")]
+            public string Key { get; set; }
+
+            [JsonProperty("value")]
+            public string Value { get; set; }
+
+            // TODO: fieldValues
```
Comment (SonarTech) [inline src/SonarQube.Client/Api/Requests/V6_30/GetPropertiesRequest.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 2 major
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/SonarQube.Client/Api/PagedRequestBase.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        private const int FirstPage = 1;
+        private const int MaximumPageSize = 500;
+
+        [JsonProperty("p")]
+        public int Page { get; set; } = FirstPage;
+
+        [JsonProperty("ps")]
+        public int PageSize { get; set; } = MaximumPageSize;
+
+        public async override Task<TResponseItem[]> InvokeAsync(HttpClient httpClient, CancellationToken token)
```
Comment (valhristov) [inline src/SonarQube.Client/Api/PagedRequestBase.cs]: There is no way to specify page range - gets pages from Page to end, but that's not needed for now

File: src/SonarQube.Client/Services/SonarQubeService2.cs [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+            where TRequest : IRequest
+        {
+            return requestFactory.Create<TRequest>(sonarQubeVersion);
+        }
+
+        public async Task ConnectAsync(ConnectionInformation connection, CancellationToken token)
+        {
+            httpClient = new HttpClient(messageHandler)
+            {
+                BaseAddress = connection.ServerUri,
```
Comment (valhristov) [inline src/SonarQube.Client/Services/SonarQubeService2.cs]: Once a request was made, BaseAddress cannot be changed anymore, hence we create a new HttpClient for each connection

File: src/Integration/Binding/BindingWorkflow.cs [hunk 9]
```diff
@@ -244,7 +244,7 @@ private ProgressStepDefinition[] CreateWorkflowSteps(IProgressController control
 
                 var roslynProfileExporter = await SafeServiceCall(
                     () => this.host.SonarQubeService.GetRoslynExportProfileAsync(qualityProfileInfo.Name, serverLanguage,
-                        cancellationToken));
+                        connectionInformation.Organization?.Key, cancellationToken));
```
Comment (Evangelink) [inline src/Integration/Binding/BindingWorkflow.cs]: Probably better to fix this in a separate commit and link it to the ticket.

File: src/SonarQube.Client/Services/SonarQubeService.cs [hunk 10]
```diff
@@ -186,93 +107,99 @@ public Uri GetProjectDashboardUrl(string projectKey)
...
+        {
+            var qualityProfiles = await InvokeRequestAsync<IGetQualityProfilesRequest, SonarQubeQualityProfile[]>(
+                request =>
+                {
+                    request.ProjectKey = projectKey;
+                    request.OrganizationKey = organizationKey;
+                },
+                token);
+
+            // TODO: add language to the request configuration above
```
Comment (SonarTech) [inline src/SonarQube.Client/Services/SonarQubeService.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by valhristov** [comment]
perhaps more tests

File: src/SonarLint.VisualStudio.Integration.sln [hunk 11]
```diff
@@ -45,6 +45,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "SonarQube.Client", "SonarQu
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "SonarQube.Client.Tests", "SonarQube.Client.Tests\SonarQube.Client.Tests.csproj", "{8FEFDB82-CBDB-4B52-B221-BB604DFD60E6}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "SonarQube.Client.RequestGenerator", "SonarQube.Client.RequestGenerator\SonarQube.Client.RequestGenerator.csproj", "{79B51C1B-7246-4862-B065-4D281BA8A384}"
```
Comment (Evangelink) [inline src/SonarLint.VisualStudio.Integration.sln]: You need to do the same for VS2017 solution.

File: src/SonarQube.Client.RequestGenerator/Properties/AssemblyInfo.cs [hunk 12]
```diff
@@ -0,0 +1,56 @@
+﻿/*
```
Comment (Evangelink) [inline src/SonarQube.Client.RequestGenerator/Properties/AssemblyInfo.cs]: I think you should remove some of the assembly properties from here and use the ones from the common assembly info.

**Comment by valhristov** [comment]
New PR was created, that's supposed to be easier to review #651

**Pull request closed without merging**